src/auth.rs

```diff
@@ -1,85 +1,139 @@
     }
 }
 
-fn get_home_dir() -> Option<PathBuf> {
-    #[cfg(target_os = "windows")]
-    if let Some(path) = env::var_os("XH_TEST_MODE_WIN_HOME_DIR") {
-        return Some(PathBuf::from(path));
-    }
-
-    home_dir()
-}
-
-fn netrc_path() -> Option<PathBuf> {
-    match env::var_os("NETRC") {
+pub fn read_netrc() -> Option<String> {
+    let netrc_path = match env::var_os("NETRC") {
         Some(path) => {
-            let pth = PathBuf::from(path);
-            if pth.exists() {
-                Some(pth)
+            let path = PathBuf::from(path);
+            if path.exists() {
+                Some(path)
             } else {
                 None
             }
         }
         None => {
-            if let Some(hd_path) = get_home_dir() {
-                [".netrc", "_netrc"]
-                    .iter()
-                    .map(|f| hd_path.join(f))
-                    .find(|p| p.exists())
-            } else {
-                None
-            }
+            let home_dir = get_home_dir()?;
+            [".netrc", "_netrc"]
+                .iter()
+                .map(|f| home_dir.join(f))
+                .find(|p| p.exists())
         }
-    }
+    }?;
+
+    fs::read_to_string(netrc_path).ok()
 }
 
-pub fn read_netrc() -> Option<String> {
-    if let Some(netrc_path) = netrc_path() {
-        if let Ok(result) = fs::read_to_string(netrc_path) {
-            return Some(result);
-        }
-    };
+pub struct DigestAuthMiddleware<'a> {
+    username: &'a str,
+    password: &'a str,
+}
 
-    None
+impl<'a> DigestAuthMiddleware<'a> {
+    pub fn new(username: &'a str, password: &'a str) -> Self {
+        DigestAuthMiddleware { username, password }
+    }
 }
 
-pub fn auth_from_netrc(machine: &str, netrc: &str) -> Option<(String, Option<String>)> {
-    if let Ok(netrc) = Netrc::parse_borrow(&netrc, false) {
-        return netrc
-            .machines
-            .into_iter()
-            .filter_map(|mach| match mach.name {
-                Some(name) if name == machine => {
-                    let user = mach.login.unwrap_or_else(|| "".to_string());
-                    Some((user, mach.password))
+impl<'a> Middleware for DigestAuthMiddleware<'a> {
+    fn handle(&mut self, mut request: Request, mut next: Next) -> Result<Response> {
+        let response = next.run(clone_request(&mut request)?)?;
+        match response.headers().get(WWW_AUTHENTICATE) {
+            Some(wwwauth) if response.status() == StatusCode::UNAUTHORIZED => {
+                let context = digest_auth::AuthContext::new(
+                    self.username,
+                    self.password,
+                    request.url().path(),
+                );
+                let mut prompt = digest_auth::parse(wwwauth.to_str()?)?;
+                let answer = prompt.respond(&context)?.to_header_string();
+                request
+                    .headers_mut()
+                    .insert(AUTHORIZATION, HeaderValue::from_str(&answer)?);
```

Comment:
I don't think it is possible to create a [mock server](https://github.com/alexliesenfeld/httpmock/issues/49) to test this code. That is unless I can modify the second request's URL during the test.